include/fiction/algorithms/physical_design/ortho_ordering_network.hpp

```diff
@@ -0,0 +1,633 @@
+
+    // If element was found
+    if (it != v.end())
+    {
+        // calculating the index
+        int64_t index = it - v.begin();
+        return index;
+    }
+    return -1;
+}
+
+/*
+ * Conditional coloring rules:
+ * 0. All 1-fan-in nodes (inverters and fan-outs) after PIs have to be colored east
+ * 1. fan-out node related to two PIs:
+ * - Case: 2-inverters -> get balanced to the beginning of the fan-out node (nl_inv has to be respected here for
+ * instance)
+ * - Case: one inverter (color inverter east and other outgoing edges south)
+ * - Case: no inverter (color one outgoing edge east and one south)
+ * 2. fan-out node related to one PI (coloring is dependent on the coloring of the whole network)
+ * 3. PI connected to node connected to PI: color connecting node east
+ */
+template <typename Ntk>
+coloring_container<Ntk> conditional_coloring(const Ntk& ntk) noexcept
+{
+    coloring_container<Ntk> ctn{ntk};
+
+    // currently viewed node
+    mockturtle::node<Ntk> current_node;
+
+#if (PROGRESS_BARS)
+    // initialize a progress bar
+    mockturtle::progress_bar bar{static_cast<uint32_t>(ctn.color_ntk.num_gates()),
+                                 "[i] determining relative positions: |{0}|"};
+#endif
+
+    // Find a coloring
+    ntk.foreach_gate_reverse(
+        [&](const auto& n, [[maybe_unused]] const auto i)
+        {
+            const auto finc = fanin_edges(ctn.color_ntk, n);
+
+            // if any incoming edge is colored east, color them all east, and south otherwise
+            const auto color =
+                std::any_of(finc.fanin_edges.cbegin(), finc.fanin_edges.cend(),
+                            [&ctn](const auto& fe) { return ctn.color_ntk.edge_color(fe) == ctn.color_east; }) ?
+                    ctn.color_east :
+                    ctn.color_south;
+
+            std::for_each(finc.fanin_edges.cbegin(), finc.fanin_edges.cend(),
+                          [&ctn, &color](const auto& fe) { recursively_paint_edges(ctn, fe, color); });
+
+            // if all incoming edges are colored east, paint the node east as well
+            if (std::all_of(finc.fanin_edges.cbegin(), finc.fanin_edges.cend(),
+                            [&ctn](const auto& fe) { return ctn.color_ntk.edge_color(fe) == ctn.color_east; }))
+            {
+                ctn.color_ntk.paint(mockturtle::node<Ntk>{n}, ctn.color_east);
+            }
+            // else, if all incoming edges are colored south, paint the node south as well
+            else if (std::all_of(finc.fanin_edges.cbegin(), finc.fanin_edges.cend(),
+                                 [&ctn](const auto& fe) { return ctn.color_ntk.edge_color(fe) == ctn.color_south; }))
+            {
+                ctn.color_ntk.paint(mockturtle::node<Ntk>{n}, ctn.color_south);
+            }
+
+#if (PROGRESS_BARS)
+            // update progress
+            bar(i);
+#endif
+        });
+
+    // Adjust only the nodes viewed in the ordering network
+    ntk.foreach_pi(
```

Comment:
This is a really long function call. Can we break it down into smaller ones for readability and testability?